src/include/txn.i

```diff
@@ -1085,17 +1088,35 @@ __wt_txn_update_check(WT_SESSION_IMPL *session, WT_UPDATE *upd)
     F_CLR(txn, WT_TXN_IGNORE_PREPARE);
     for (; upd != NULL && !__wt_txn_upd_visible(session, upd); upd = upd->next) {
         if (upd->txnid != WT_TXN_ABORTED) {
-            if (ignore_prepare_set)
-                F_SET(txn, WT_TXN_IGNORE_PREPARE);
-            WT_STAT_CONN_INCR(session, txn_update_conflict);
-            WT_STAT_DATA_INCR(session, txn_update_conflict);
-            return (__wt_txn_rollback_required(session, "conflict between concurrent operations"));
+            rollback = true;
+            break;
         }
     }
 
+    /*
+     * Check conflict against the on page value if there is no update on the update chain except
+     * aborted updates. Otherwise, we would have either already detected a conflict if we saw an
+     * uncommitted update or determined that it would be safe to write if we saw a committed update.
+     */
+    if (upd == NULL && cbt != NULL && cbt->btree->type != BTREE_COL_FIX && cbt->ins == NULL) {
```

Comment:
Does this need a `!rollback` check in the clause as well? I see that you assert that it's not already chosen a rollback - does that mean upd should be NULL if rollback is set? That seems a bit fragile, and I think this code would be better if it handles the unexpected case - asserts are only active in WT_HAVE_DIAGNOSTIC builds, so being double safe seems like the right choice here.

Reply:
Yeah, makes sense. Fixed